src/sleep/sleep.c

```diff
@@ -263,40 +263,94 @@ static int execute(
+
+                r = execute(sleep_config, SLEEP_SUSPEND, NULL);
+                if (r < 0)
+                        return r;
+
+                r = fd_wait_for_event(tfd, POLLIN, 0);
+                if (r < 0)
+                        return log_error_errno(r, "Error polling timerfd: %m");
+
+                if (!FLAGS_SET(r, POLLIN))  { /* We woke up before the alarm time, calculate battery discharged and estimate rate. */
+                        after_timestamp = now(CLOCK_BOOTTIME);
+
+                        time_elapsed = (after_timestamp - before_timestamp) / USEC_PER_SEC;
 
-        tfd = safe_close(tfd);
+                        current_capacity = read_battery_capacity_percentage();
+                        if (current_capacity < 0)
+                                return log_error_errno(errno, "Error fetching battery capacity percentage: %m");
+
+                        log_debug("%d%% battery discharged in %f seconds time",
+                                        last_capacity - current_capacity,
+                                        time_elapsed);
+
+                        estimated_discharge_rate = ((last_capacity - current_capacity) * 60 * 60) / time_elapsed;
+
+                        log_debug("Battery discharge rate is %d%% per hour", estimated_discharge_rate);
+
+                        if (estimated_discharge_rate > 0 && estimated_discharge_rate < 200) {
+                                r = put_battery_discharge_rate(estimated_discharge_rate);
+                                if (r < 0)
+                                        log_warning_errno(r, "Failed to update battery discharge rate: ignoring %m");
+                        }
+                        return 0;
+                }
+
+                log_debug("Attempting to estimate battery after waking from %s hours sleep timer",
+                          FORMAT_TIMESPAN(suspend_interval, USEC_PER_HOUR));
+
+                /* If woken up after alarm time, calculate battery capacity and battery capacity_level and estimate discharge rate */
+                current_capacity = read_battery_capacity_percentage();
+
+                if (current_capacity < 0)
+                        return log_error_errno(errno, "Error fetching battery capacity: %m");
+
+                estimated_discharge_rate = (last_capacity - current_capacity) / (suspend_interval * USEC_PER_HOUR);
+
+                log_debug("Battery discharge rate is %d%% per hour", estimated_discharge_rate);
+
+                if (estimated_discharge_rate > 0 && estimated_discharge_rate < 200) {
+                        r = put_battery_discharge_rate(estimated_discharge_rate);
+                        if (r < 0)
+                                log_warning_errno(r, "Failed to update battery discharge rate, ignoring: %m");
+                }
 
-        /* If woken up after alarm time, hibernate */
-        log_debug("Attempting to hibernate after waking from %s timer",
```

Comment:
this log message should be kept as it's useful, just changed to remove the mention of the timer

Reply:
log_debug("Attempting to estimate battery after waking from %s hours sleep timer",
                          FORMAT_TIMESPAN(suspend_interval, USEC_PER_HOUR));

                /* If woken up after alarm time, calculate battery capacity and battery capacity_level and estimate discharge rate */
                current_capacity = read_battery_capacity_percentage();

It is there. I just updated it.

Reply:
Yeah but that's a different message, and it's not before hibernate - it's good to log just before going into hibernation, and it should be left in the same position, just without mentioning the timer as it's no longer relevant